src/mesh/msh_gmshreader.cc

```diff
@@ -291,48 +289,44 @@ void GmshMeshReader::read_data_header(GMSH_DataHeader &head) {
 
 
 template<typename T>
-typename ElementDataCache<T>::ComponentDataPtr GmshMeshReader::get_element_data( GMSH_DataHeader &search_header,
-		std::vector<int> const & el_ids, unsigned int component_idx)
+typename ElementDataCache<T>::ComponentDataPtr GmshMeshReader::get_element_data( std::string field_name, double time,
+		unsigned int n_entities, unsigned int n_components, bool &actual, std::vector<int> const & el_ids, unsigned int component_idx)
 {
     using namespace boost;
+    //
 
-    GMSH_DataHeader actual_header = find_header(search_header.time, search_header.field_name);
-    if ( !current_cache_->is_actual(actual_header.time, search_header.field_name) ) {
+    GMSH_DataHeader actual_header = find_header(time, field_name);
+    if ( !current_cache_->is_actual(actual_header.time, field_name) ) {
 
```

Comment:
Only the base class will have templated get_element_data method, this method creates CacheData<T>  
entry and pass it as CacheDataBase (non templated , similar to OutputDataBase) to internal reader method
read_data. This method is virtual with separate implementation for GMSH and VTK (and separate from ascii and binary). These methods call virtual methods of CacheData<T> to read individual data fields from ascii or binary streams.
1. move check of actual current_cache_ into base class
2. In base class create new ElementDataCache<T>